### PR TITLE
Add complex text layout support

### PR DESCRIPTION
**Pull request by asmaAL-Bahanta**
This pull request adds support for languages that require [complex text
layout](https://en.wikipedia.org/wiki/Complex_text_layout).

We are using [libraqm](https://github.com/HOST-Oman/libraqm), a small source
code-only library that wraps FriBidi (for bidirectional text support) and
HarfBuzz (for text shaping), and does proper BiDi and script itemization.

The CTL support is enabled by default but can be disabled at compiling time,
and we provide a fallback function that uses your original code without CTL
support.

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -208,6 +208,31 @@ if test "$gd_with_LIBFREETYPE" = "yes"; then
   AC_DEFINE([HAVE_FT2BUILD_H], [1], [Define if you have the ft2build.h header.])
 fi
 
+dnl CTL configure tests
```
Comment (vapier) [inline configure.ac]: delete all of this and use the new GD_LIB_PKG_CHECK helper

**Comment by vapier** [comment]
this is against an old version of gd, so please rebase onto the latest master, and you must make sure Travis is passing first

File: configure.ac (hunk 1 above)
Comment (asmaAL-Bahanta) [inline configure.ac]: we are using `AC_ARG_ENABLE` because ctl is a feature not a single package, so it should be `--enable-ctl` as `--with-ctl` does not seem appropriate.

**Comment by asmaAL-Bahanta** [comment]
We fixed the failing test and travis build on linux.
Mac build uses cmake but we couldn't get it working under Ubuntu 14.04 so we didn't modify it to check for ctl libraries.

File: configure.ac [hunk 2]
```diff
@@ -208,6 +208,31 @@ if test "$gd_with_LIBFREETYPE" = "yes"; then
...
+AC_ARG_ENABLE(ctl,
+    [AS_HELP_STRING([--disable-ctl],
+                    [disable complex text layout support])],
+    [enable_ctl=$enableval],
```
Comment (vapier) [inline configure.ac]: use the variable name gd_enable_CTL to match the existing gd_with_XXX style

File: configure.ac [hunk 3]
```diff
@@ -208,6 +208,31 @@ if test "$gd_with_LIBFREETYPE" = "yes"; then
...
+    [enable_ctl='yes'])
```
Comment (vapier) [inline configure.ac]: this should be "auto", not "yes"

File: configure.ac [hunk 4]
```diff
@@ -208,6 +208,31 @@ if test "$gd_with_LIBFREETYPE" = "yes"; then
...
+
+if test "x$gd_with_LIBFREETYPE" != "xyes" -a "x$enable_ctl" != 'xno'; then
+    AC_MSG_ERROR([Complex text support requested but FreeType is missing])
+fi
+
+have_ctl=""
+if test "x$enable_ctl" != 'xno'; then
+    PKG_CHECK_MODULES(CTL, [fribidi harfbuzz], have_ctl="yes", [])
```
Comment (vapier) [inline configure.ac]: split this up into two lib checks and use standard vars:
``PKG_CHECK_MODULES([FRIBIDI], [fribidi], [gd_with_FRIBIDI=yes], [gd_with_FRIBIDI=no])``

then you can do:

    if test "x$gd_with_FRIBIDI$gd_with_HARFBUZZ" != "xyesyes"; then
      if test "x$gd_enable_CTL" = "xyes"; then
        AC_MSG_ERROR(...)
      fi
      gd_enable_CTL="no"
    fi

File: configure.ac [hunk 5]
```diff
@@ -208,6 +208,31 @@ if test "$gd_with_LIBFREETYPE" = "yes"; then
...
+fi
+
+if test "x$have_ctl" == "xyes"; then
```
Comment (vapier) [inline configure.ac]: this code is invalid -- you can't use ``==`` w/test.  you must use ``=``.

File: configure.ac [hunk 6]
```diff
@@ -208,6 +208,31 @@ if test "$gd_with_LIBFREETYPE" = "yes"; then
...
+    AC_DEFINE(HAVE_CTL, 1, [Define if you have the complex text support.])
+    CPPFLAGS="$CPPFLAGS $CTL_CFLAGS"
+    LIBS="$LIBS $CTL_LIBS"
+    FEATURES="GD_CTL $FEATURES"
+    AC_CHECK_FUNC(fribidi_reorder_runs, AC_DEFINE(HAVE_FRIBIDI_REORDER_RUNS,  1, [Define if FriBiDi has fribidi_reorder_runs]), [])
+fi
+AM_CONDITIONAL(HAVE_CTL, test "x$have_ctl" == "xyes")
```
Comment (vapier) [inline configure.ac]: same error here w/``==``

File: src/gdft.c [hunk 7]
```diff
@@ -431,6 +431,66 @@ gdTcl_UtfToUniChar (char *str, Tcl_UniChar * chPtr)
...
+#else
+typedef struct {
+	int index;
+	int x_offset;
+	int x_advance;
+	int y_offset;
+	uint32_t cluster;
+} raqm_glyph_info_t;
+
+unsigned fallback_shape(unsigned int *text,int len,
```
Comment (vapier) [inline src/gdft.c]: put a space after the first ``,``

Comment (vapier) [inline src/gdft.c]: this should be marked static

File: src/raqm.c [hunk 8]
```diff
@@ -0,0 +1,695 @@
...
+    hb_tag_to_string (hb_script_to_iso15924_tag (script), buff); \
+    buff[4] = '\0';
+#endif
+
+/* for older fribidi versions */
+#ifndef HAVE_FRIBIDI_REORDER_RUNS
+typedef struct _FriBidiRunStruct FriBidiRun;
+struct _FriBidiRunStruct
+{
+  FriBidiRun *prev;
```
Comment (vapier) [inline src/raqm.c]: fix the style in this file -- indent with tabs

File: src/raqm.c [hunk 9]
```diff
@@ -0,0 +1,695 @@
...
+};
+
+/* Reverse an array of runs */
+static void reverse_run (FriBidiRun *arr, const FriBidiStrIndex len)
+{
+  FriBidiStrIndex i;
+
+  assert (arr);
+
+  for (i = 0; i < len/2; i++)
```
Comment (vapier) [inline src/raqm.c]: use spaces around the ``/`` operator

File: src/raqm.c [hunk 10]
```diff
@@ -0,0 +1,695 @@
...
+  FriBidiStrIndex i;
+  FriBidiLevel level;
+  FriBidiLevel last_level = -1;
+  FriBidiLevel max_level = 0;
+  FriBidiStrIndex run_count = 0;
+  FriBidiStrIndex run_start = 0;
+  FriBidiStrIndex run_index = 0;
+
+ if
+    (len == 0)
```
Comment (vapier) [inline src/raqm.c]: code style here is broken

File: src/raqm.c [hunk 11]
```diff
@@ -0,0 +1,695 @@
...
+    {
+      goto out;
+    }
+
+  assert (bidi_types);
+  assert (embedding_levels);
+
+  {
```
Comment (vapier) [inline src/raqm.c]: use of braces here is pointless

File: src/raqm.h [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+ * SOFTWARE.
+ *
+ */
+
+#ifndef _RAQM_H_
+#define _RAQM_H_
+
+/* For enabling debug mode */
+/*#define DEBUG 1*/
+#ifdef DEBUG
```
Comment (vapier) [inline src/raqm.h]: i really don't want to see these debug/test macros being duplicated further.  add some to gdhelpers.h and update the existing GD2_DBG/IO_DBG/GD_SS_DBG defines in the code base to use those.

Comment (asmaAL-Bahanta) [inline src/raqm.h]: These  macros are used for  [libraqm] (https://github.com/HOST-Oman/libraqm) tests.

File: src/raqm.c (hunk 8 above)
Comment (asmaAL-Bahanta) [inline src/raqm.c]: This code is part of [libraqm](https://github.com/HOST-Oman/libraqm), and it is indented by spaces.
This particular code was submitted to FriBiDi and uses it's code style to be able to  sync it easily.

**Comment by asmaAL-Bahanta** [comment]
I addressed all other comments.

**Comment by Fahad-Alsaidi** [comment]
any update?

**Comment by vapier** [comment]
i don't think we want to bundle libraqm in gd itself.  it should be treated as any other external 3rd party library.

**Comment by asmaAL-Bahanta** [comment]
I have updated the pull request with raqm as standalone library.

File: src/gdft.c [hunk 13]
```diff
@@ -1276,22 +1321,89 @@ BGD_DECLARE(char *) gdImageStringFTEx (gdImage * im, int *brect, int fg, char *f
...
-		glyph_index = FT_Get_Char_Index (face, ch);
+	FT_Activate_Size (platform_independent);
+#ifdef HAVE_LIBRAQM
+	raqm_t *rq;
+	raqm_direction_t dir;
+	dir = RAQM_DIRECTION_DEFAULT;
+	rq = raqm_create ();
+	if(rq == NULL){
+		raqm_destroy(rq);
+		if (tmpstr)
```
Comment (vapier) [inline src/gdft.c]: no need to do NULL checks

File: src/gdft.c [hunk 14]
```diff
@@ -1276,22 +1321,89 @@ BGD_DECLARE(char *) gdImageStringFTEx (gdImage * im, int *brect, int fg, char *f
...
+			gdFree (tmpstr);
+		gdCacheDelete (tc_cache);
+		gdMutexUnlock (gdFontCacheMutex);
+		return "Problem creating Raqm Object";
+	}
+	if (!raqm_set_text (rq, text, i)){
+		raqm_destroy(rq);
+		if (tmpstr)
```
Comment (vapier) [inline src/gdft.c]: no need to do NULL checks

File: src/gdft.c [hunk 15]
```diff
@@ -1276,22 +1321,89 @@ BGD_DECLARE(char *) gdImageStringFTEx (gdImage * im, int *brect, int fg, char *f
...
+			gdFree (tmpstr);
+		gdCacheDelete (tc_cache);
+		gdMutexUnlock (gdFontCacheMutex);
+		return "Problem setting Raqm Text";
+	}
+	if (!raqm_set_freetype_face (rq, face)){
+		if (tmpstr)
```
Comment (vapier) [inline src/gdft.c]: no need to do NULL checks

File: src/gdft.c [hunk 16]
```diff
@@ -1276,22 +1321,89 @@ BGD_DECLARE(char *) gdImageStringFTEx (gdImage * im, int *brect, int fg, char *f
...
 
-		/* Convert character code to glyph index */
-		glyph_index = FT_Get_Char_Index (face, ch);
+	FT_Activate_Size (platform_independent);
+#ifdef HAVE_LIBRAQM
+	raqm_t *rq;
+	raqm_direction_t dir;
+	dir = RAQM_DIRECTION_DEFAULT;
+	rq = raqm_create ();
+	if(rq == NULL){
```
Comment (vapier) [inline src/gdft.c]: space after the if

File: src/gdft.c [hunk 17]
```diff
@@ -431,6 +431,68 @@ gdTcl_UtfToUniChar (char *str, Tcl_UniChar * chPtr)
...
+		int y_advance;
+		int x_offset;
+		int y_offset;
+		uint32_t cluster;
+} raqm_glyph_t;
+
+static unsigned
+fallback_shape(unsigned int *text, int len,
+		FT_Face face, gdFTStringExtraPtr strex,
+		raqm_glyph_t** glyph_info)
```
Comment (vapier) [inline src/gdft.c]: spaces go before the ** not after

File: src/gdft.c [hunk 18]
```diff
@@ -431,6 +431,68 @@ gdTcl_UtfToUniChar (char *str, Tcl_UniChar * chPtr)
 	return 1;
 }
 
+#ifdef HAVE_LIBRAQM
+#include <raqm.h>
+#else
+typedef struct {
+	unsigned int index;
+		int x_advance;
```
Comment (vapier) [inline src/gdft.c]: indentation is broken here

also, you use int, but you store values that are FT_Pos (signed long), so perhaps these should all be FT_Pos ?

File: src/gdft.c [hunk 19]
```diff
@@ -1276,22 +1321,89 @@ BGD_DECLARE(char *) gdImageStringFTEx (gdImage * im, int *brect, int fg, char *f
...
-		if ( ! (strex && (strex->flags & gdFTEX_DISABLE_KERNING))
-		        && ! FT_IS_FIXED_WIDTH(face)
-		        && FT_HAS_KERNING(face)
-		        && previous
-		        && glyph_index)
-			FT_Get_Kerning (face, previous, glyph_index, ft_kerning_default, &delta);
-		else
-			delta.x = delta.y = 0;
+	if (!raqm_set_par_direction (rq, dir)){
+		if (tmpstr)
```
Comment (vapier) [inline src/gdft.c]: no need to do NULL checks

File: src/gdft.c [hunk 20]
```diff
@@ -1276,22 +1321,89 @@ BGD_DECLARE(char *) gdImageStringFTEx (gdImage * im, int *brect, int fg, char *f
...
+	if (!raqm_layout (rq)){
+		if (tmpstr)
+			gdFree (tmpstr);
+		gdCacheDelete (tc_cache);
+		gdMutexUnlock (gdFontCacheMutex);
+		return "Problem setting Raqm layout";
+	}
+	info = raqm_get_glyphs (rq, &count);
+	if (info == NULL){
+		if (tmpstr)
```
Comment (vapier) [inline src/gdft.c]: no need to do NULL checks

File: .travis.yml [hunk 21]
```diff
@@ -11,27 +11,33 @@ os:
   - linux
   - osx
 
+sudo: required # needed for trusty
+dist: trusty # needed for Raqm
+
 before_install:
   - if [ "$TRAVIS_OS_NAME" = "linux" ];
     then
+      sudo apt-add-repository -y ppa:as-bahanta/raqm;
```
Comment (vapier) [inline .travis.yml]: i would leave this out which means raqm would not be tested by default.  you should submit the library for inclusion into places first.

Comment (khaledhosny) [inline .travis.yml]: May be it is better to use a matrix and build once with and once without raqm, that would make sure the Raqm path does not bitrot? (note that the Mac build doesn’t use Raqm, so currently both code paths are exercised, though on different platforms).

Comment (khaledhosny) [inline .travis.yml]: Also even if Raqm was added to Debian and Ubuntu today, it would be years before it reaches Travis, so the PPA will be needed for sometime anyway.

Comment (vapier) [inline .travis.yml]: i'm aware of the impact.  it's unfortunate.  but i don't think we want to make gd's test system dependent upon a random ppa.  please don't take this personally as i would have the same response for anyone who wanted to add a ppa to the list.

Comment (khaledhosny) [inline .travis.yml]: So what are the alternatives? I fear that the code will easily bitrot if it is not actively built and tested.

Comment (vapier) [inline .travis.yml]: your concern is certainly valid.  but i think that's the route we should take as long as the package is unavailable in the distro.

looking at the changes, i don't think it's that big of a deal as code is not that intrusive.

Comment (khaledhosny) [inline .travis.yml]: OK, I think this should work.

**Comment by asmaAL-Bahanta** [comment]
I have Modified the pull request so that Raqm will not be tested by default.

File: .travis.yml [hunk 22]
```diff
@@ -11,27 +11,27 @@ os:
   - linux
   - osx
 
+
```
Comment (vapier) [inline .travis.yml]: please avoid unrelated whitespace changes

File: src/gdft.c [hunk 23]
```diff
@@ -431,6 +431,68 @@ gdTcl_UtfToUniChar (char *str, Tcl_UniChar * chPtr)
...
+		FT_Face face, gdFTStringExtraPtr strex,
+		raqm_glyph_t **glyph_info)
+{
+	FT_UInt glyph_index, previous;
+	FT_Vector delta;
+	FT_Error err;
+	int count;
+	glyph_index = 0;
+	previous = 0;
+	raqm_glyph_t *g_info = (raqm_glyph_t*) gdMalloc(sizeof(raqm_glyph_t) * len + 1);
```
Comment (vapier) [inline src/gdft.c]: you need to check the return value to make sure it's not NULL

File: src/gdft.c [hunk 24]
```diff
@@ -950,20 +1012,23 @@ BGD_DECLARE(char *) gdImageStringFTEx (gdImage * im, int *brect, int fg, char *f
...
 	double sin_a = sin (angle);
 	double cos_a = cos (angle);
-	int len, i, ch;
+	int  i, ch;
 	font_t *font;
 	fontkey_t fontkey;
 	char *next;
 	char *tmpstr = 0;
+	unsigned int *text;
+	raqm_glyph_t *info = NULL;
```
Comment (vapier) [inline src/gdft.c]: this is used outside of #ifdef HAVE_LIBRAQM, so won't this fail ?

File: src/gdft.c [hunk 25]
```diff
@@ -1276,22 +1321,83 @@ BGD_DECLARE(char *) gdImageStringFTEx (gdImage * im, int *brect, int fg, char *f
...
+		gdMutexUnlock (gdFontCacheMutex);
+		return "Problem getting Raqm glyphs info";
+	}
+#else
+	count = fallback_shape (text , i, face, strex, &info);
+#endif
+
+	gdFree (text);
+
+	for(i = 0; i< count; i++) {
```
Comment (vapier) [inline src/gdft.c]: space before the <

File: src/gdft.c (hunk 24 above)
Comment (asmaAL-Bahanta) [inline src/gdft.c]: No it won't fail, In case raqm is not defined there is a struct called `raqm_glyph_t` will be defined instead.

**Comment by asmaAL-Bahanta** [comment]
Updated the pull request.

File: src/gdft.c (hunk 23 above)
Comment (asmaAL-Bahanta) [inline src/gdft.c]: I have added Null check.

File: src/gdft.c [hunk 26]
```diff
@@ -431,6 +431,71 @@ gdTcl_UtfToUniChar (char *str, Tcl_UniChar * chPtr)
...
+static unsigned
+fallback_shape(unsigned int *text, int len,
+		FT_Face face, gdFTStringExtraPtr strex,
+		raqm_glyph_t **glyph_info)
+{
+	FT_UInt glyph_index, previous;
+	FT_Vector delta;
+	FT_Error err;
+	int count;
+	glyph_index = 0;
```
Comment (vapier) [inline src/gdft.c]: just inline these two assignments above where the vars are declared

File: src/gdft.c [hunk 27]
```diff
@@ -431,6 +431,71 @@ gdTcl_UtfToUniChar (char *str, Tcl_UniChar * chPtr)
...
+	previous = 0;
+	raqm_glyph_t *g_info = (raqm_glyph_t*) gdMalloc(sizeof(raqm_glyph_t) * len + 1);
+	if (!g_info) {
+		return NULL;
```
Comment (vapier) [inline src/gdft.c]: this func has a return type of unsigned, so using a pointer here makes no sense

File: src/gdft.c [hunk 28]
```diff
@@ -431,6 +431,71 @@ gdTcl_UtfToUniChar (char *str, Tcl_UniChar * chPtr)
...
+			&& previous
+			&& glyph_index)
+			FT_Get_Kerning (face, previous, glyph_index, ft_kerning_default, &delta);
+		else
+			delta.x = delta.y = 0;
+
+		err = FT_Load_Glyph (face, glyph_index, FT_LOAD_DEFAULT);
+		if (err) {
+			gdMutexUnlock (gdFontCacheMutex);
+			return 0;
```
Comment (vapier) [inline src/gdft.c]: doesn't this leak the g_info memory ?

File: src/gdft.c [hunk 29]
```diff
@@ -1276,22 +1324,83 @@ BGD_DECLARE(char *) gdImageStringFTEx (gdImage * im, int *brect, int fg, char *f
...
+	}
+	if (!raqm_set_text (rq, text, i)) {
+		raqm_destroy(rq);
+		gdFree (tmpstr);
+		gdCacheDelete (tc_cache);
+		gdMutexUnlock (gdFontCacheMutex);
+		return "Problem setting Raqm Text";
+	}
+	if (!raqm_set_freetype_face (rq, face)) {
+		gdFree (tmpstr);
```
Comment (vapier) [inline src/gdft.c]: shouldn't this raqm_destry too ?

File: src/gdft.c [hunk 30]
```diff
@@ -1276,22 +1324,83 @@ BGD_DECLARE(char *) gdImageStringFTEx (gdImage * im, int *brect, int fg, char *f
...
-		/* Convert character code to glyph index */
-		glyph_index = FT_Get_Char_Index (face, ch);
+	FT_Activate_Size (platform_independent);
+#ifdef HAVE_LIBRAQM
+	raqm_t *rq;
+	raqm_direction_t dir;
+	dir = RAQM_DIRECTION_DEFAULT;
+	rq = raqm_create ();
+	if (rq == NULL) {
+		raqm_destroy(rq);
```
Comment (vapier) [inline src/gdft.c]: style here says space before the (

File: src/gdft.c [hunk 31]
```diff
@@ -1276,22 +1324,83 @@ BGD_DECLARE(char *) gdImageStringFTEx (gdImage * im, int *brect, int fg, char *f
...
+		gdFree (tmpstr);
+		gdCacheDelete (tc_cache);
+		gdMutexUnlock (gdFontCacheMutex);
+		return "Problem creating Raqm Object";
+	}
+	if (!raqm_set_text (rq, text, i)) {
+		raqm_destroy(rq);
```
Comment (vapier) [inline src/gdft.c]: style here says space before the (

File: src/gdft.c [hunk 32]
```diff
@@ -1276,22 +1324,83 @@ BGD_DECLARE(char *) gdImageStringFTEx (gdImage * im, int *brect, int fg, char *f
...
+		gdMutexUnlock (gdFontCacheMutex);
+		return "Problem getting Raqm glyphs info";
+	}
+#else
+	count = fallback_shape (text , i, face, strex, &info);
+#endif
+
+	gdFree (text);
+
+	for(i = 0; i < count; i++) {
```
Comment (vapier) [inline src/gdft.c]: space before the (

**Comment by asmaAL-Bahanta** [comment]
I have addressed all the comments.

File: src/gdft.c [hunk 33]
```diff
@@ -431,6 +431,70 @@ gdTcl_UtfToUniChar (char *str, Tcl_UniChar * chPtr)
...
+		FT_Face face, gdFTStringExtraPtr strex,
+		raqm_glyph_t **glyph_info)
+{
+	FT_UInt glyph_index = 0, previous = 0;
+	FT_Vector delta;
+	FT_Error err;
+	int count;
+	raqm_glyph_t *g_info = (raqm_glyph_t*) gdMalloc(sizeof(raqm_glyph_t) * len + 1);
+	if (!g_info) {
+		return NULL;
```
Comment (vapier) [inline src/gdft.c]: return value is int now, but you're still returning NULL -- a pointer

File: src/gdft.c [hunk 34]
```diff
@@ -431,6 +431,70 @@ gdTcl_UtfToUniChar (char *str, Tcl_UniChar * chPtr)
...
+			&& FT_HAS_KERNING(face)
+			&& previous
+			&& glyph_index)
+			FT_Get_Kerning (face, previous, glyph_index, ft_kerning_default, &delta);
+		else
+			delta.x = delta.y = 0;
+
+		err = FT_Load_Glyph (face, glyph_index, FT_LOAD_DEFAULT);
+		if (err) {
+			gdMutexUnlock (gdFontCacheMutex);
```
Comment (vapier) [inline src/gdft.c]: is this supposed to mess with the mutex ?  seems wrong to me.  this should really be in the caller's code paths and not here.

**Comment by asmaAL-Bahanta** [comment]
Updated.

**Comment by asmaAL-Bahanta** [comment]
Any updates ?!

**Comment by khaledhosny** [comment]
Any thing remaining to make the pull request ready?

**Comment by khaledhosny** [comment]
Is there any chance of this code being merged?

**Comment by vapier** [comment]
please rebase onto master and discard the travis commit ... that's been fixed since.  i'll try and take a look after that.

**Comment by pierrejoye** [comment]
Have you checked gd-pango? Or what do you miss there?

I would also be sure that libraqm is portable (incl Windows), do you have any info if there is work in progress for that?

**Comment by vapier** [comment]
merged via #246, thanks!

**Comment by pierrejoye** [comment]
Hm.

Why is it merged?

**Comment by vapier** [comment]
you mean this PR, or libraqm support in general ?  i thought we were moving towards supporting this, but i just saw/remembered your gd-pango question.

@khalidalqassabi: can you answer the gd-pango part ?

**Comment by pierrejoye** [comment]
I am not confident to support it inside libgd as of now. We can have an extra module like what we did with pango until this library is proven to be stable and portable. The amount of changes adds complexity to something I would rather keep simple.

I also asked what is the core advantage of this library compared to pango f.e. that's something I like yo clarify before we decide to maintain support for this library.

Not really the biggest problem but these changes are creating issues with a fix we have to merge.

**Comment by pierrejoye** [comment]
The raqm support in general :)

I am not saying I do not like it but it is a bit early to consider it as
core feature.

**Comment by vapier** [comment]
yeah, sorry about that.  if they don't follow up, i'll revert.  fortunately we managed to delete a lot of code before it was merged ;).

**Comment by khaledhosny** [comment]
The main benefit is less dependencies, Pango depends on GLib which can be pain to build on some platforms and almost depends on Cairo (it has alternative backends but Cairo is the most developed one). Also Raqm is much smaller than Pango (though it does a bit less than Pango as well).

**Comment by khaledhosny** [comment]
BTW, the original intent was to not have a separate libraqm at all (it is just a single C file that could have been part of the repository).

**Comment by Fahad-Alsaidi** [comment]
@pierrejoye  does anybody work in gd-pango?? last commit was in 2008 !! 

However, Raqm & pango do the same thing but Raqm is lighter and only depends on Harfbuzz & FriBiDi. I really don't care what will you prefer in GD but I only wish you could add CTL support in GD before 2050 !!
